rfcs/002 - Database Access.md

```diff
@@ -0,0 +1,160 @@
+
+It would still require us to connect to the URL/database/etc, to gather information to provide to the LLM, but it would be a more modular "document generator" mechanism that could be plugin-based where parsing of the markdown in the description section of the function blocks would figure out which document(s) to generate and provide to the LLM during function generation.
+
+This reduces the amount of new syntax dramatically, but eschews any database schema management within Marsha for at least the time being.
+
+Some of the desired functionality, like a shared connection object to reduce connection pressure on the database, could be done with a Marsha function told to memoize generated connection objects and other Marsha functions told to acquire the connection from said function. This places the burden on the user to "do the right thing" but keeps language simplicity and flexibility high. Potentially the user burden can be reduced by developing a standard library for Marsha and/or a module system.
+
+```md
+# func get_users(): list of user objects
+
+This function gets all user records from [the database][1] and returns them, or fails if unable to connect to the database.
+
+* get_users() = [{"name": "John Doe", "age": 27, "enjoys_udon": true}, {"name": "Jane Doe", "age": 28, "enjoys_udon": false}]
+* get_users() raises database access exception
+
+# references
+
+[1]: ./db.sqlite
+```
+
+In this approach, *all* of the logic SQL Pal has to guess what tables are relevant will be necessary (or we make it very expensive with large-context GPT4 calls).
+
+#### Extensions Syntax
+
+LLM abilities probably haven't plateaued yet, and I see value in both the Fully-separated syntax and the References syntax -- which is desired may depend on personal preference more than anything, but needing to choose one over the other may limit the future of Marsha in undesirable ways. At least while we're still figuring out what to leave to the LLM, what to guide the LLM with, and what to explicitly require the user to provide, it could be useful to have users specify which manipulations of the Markdown syntax they even want to have with special interpreter directives defined at the beginning of the file, kinda like [Raku](https://www.raku.org/) or [Racket](https://racket-lang.org/).
+
+We could define a `# using {extension 1} [, {extension 2} ...]` header that *must* be the first if present, specifying what parsing extensions are to be used on the text after it. We could even potentially put *all* functionality behind different extensions to give us the flexibility to make breaking changes to things like functions without breaking any code still using the older function behavior, but that could be a wonky barrier to entry for people who are not programmers.
```

Comment:
yes, when I started reading it my thought was that las sentence, but thinking about it, while we get a "final" solution users that are not programmers either are not going to use these "advance modes" so often or they have not discovered marsha yet?

Reply:
While we're actively developing on it during this alpha state I think that's fine, but I do think a default "base" extension that's automatically used and extensions long-term are just development artifacts and/or for esoteric purposes in the future. I think I mention something like that in the following paragraph.